### DISCUSSION TITLE
RFC: Associating attributes with graphs

### DISCUSSION THREAD

**Discussion by avdgrinten** [networkit/networkit#681] [Ideas]
During ND'20, there were some requests about better support for attributed graphs. As far as I can see, there are three interesting kinds of attributes: (i) graph attributes, (ii) node attributes, and (iii) edge attributes.

To efficiently implement node and edge attributes, we should index them by their node / edge ID. This means that edge attributes will only work on graphs that have edge IDs. Using this method, it's easy to implement scalar attributes (bools, integers, doubles, strings).

Regarding the syntax, we have a few options. In C++, we can use tag dispatch to differentiate between different attributes types (which should be statically typed, for performance reasons). In Python, there is no tag dispatch but everything is dynamically typed anyway, so we can keep the same syntax.

```cpp
// Attach a new node attribute to the graph.
auto colors = G.nodeAttribs().attach(NetworKit::intAttrib, "color");
// Retrieve a reference to an existing node attribute.
auto colors = G.nodeAttribs().access(NetworKit::intAttrib, "color");

// Getting and settings attributes for node n:
// We could probably also overload operator[] here.
std::cout << colors.get(n) << std::endl;
colors.set(n, 42);
```

Open questions:
* Does this approach make sense?
* Does the syntax make sense?
* Can this be made fast enough? For example, can `addNode()` / `addEdge()` still be implemented efficiently? We should probably avoid doing additional work in these functions; instead, we can update the attribute data structures lazily. However, remapping of nodes / edges needs to be adjusted.
* Should attributes be identified by strings, as proposed above?
* Do we need a more ergonomic API for graph attributes (where there is no index involved)?
* Do we want some `std::any`-typed attribute to handle non-scalar attributes? Or are non-scalar attributes out of scope for now?
* How is ownership of attributes managed? It probably makes sense to let the graph own the attributes (instead of sharing one "column" of attributes among multiple graphs). That way, we do not run into nasty concurrency issues.

**Comment by hmeyerhenke** [comment]
I think this is quite close to what the team back then had in mind when they introduced edge indexing. The rationale was that the user code was taking care of attributes. But what you propose seems to be a reasonable move to unify this. Performance should be kept in mind. And a simple (yet extensible) solution is probably better in the beginning than a solution that handles all corner cases (but would then burden our resources too much). I leave more detailed comments to those programming more with NetworKit.

**Comment by j6k4m8** [comment]
I wanted to mention that I was among the users at ND20 that made the request for node/edge/graph attributes. Since, I have written [Grand](https://github.com/aplbrain/grand), a Python library that serves as an abstraction layer for graph libraries (so an algorithm written in NetworkX can be run on a Networkit graph, for example).

I'd be happy to offer some insight here if helpful to the team, but one important note is that several graph libs I've used (in particular, Petgraph in Rust and a few others in Go) have an attributed version as well as an unattributed version of their `Graph` interfaces, which has the benefit of leaving performance alone for the "base" unattributed case, but enabling a wider set of algorithms on a more memory-hungry attributed version.

**Reply by hmeyerhenke** [reply]
Jordan, we'd be happy to get your insights and will get back to you when this really starts -- hopefully soon.

**Reply by j6k4m8** [reply]
Awesome!! I'd love to share lessons-learned and be helpful to you in any way I can!
